### PR TITLE
br, lightning: unify the calling of WaitForScatterRegion

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref #51533

Problem Summary:

### What changed and how does it work?

Change at BR side: If the scattered number is changed which means scatter progress move forward, don't increase the retry counter. This may enlarge the total wait time.

Change at lightning side: When found a region needs to be scattered again, save it and **batch** scatter all needed regions in this round, rather than scatter it immediately.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
  - lightnign CI https://do.pingcap.net/jenkins/blue/organizations/jenkins/pingcap%2Ftidb%2Fpull_lightning_integration_test/detail/pull_lightning_integration_test/94/pipeline 
  - BR CI https://do.pingcap.net/jenkins/blue/organizations/jenkins/pingcap%2Ftidb%2Fpull_br_integration_test/detail/pull_br_integration_test/66/pipeline
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/restore/split/client.go [hunk 1]
```diff
@@ -636,13 +634,113 @@ func (c *pdClient) IsScatterRegionFinished(
...
+
+		if len(needRescatter) > 0 {
+			scatterErr := c.ScatterRegions(ctx, needRescatter)
+			if scatterErr != nil {
+				if !common.IsRetryableError(scatterErr) {
+					retErr = scatterErr
+					return nil
+				}
+				if retErr == nil {
+					retErr = scatterErr
```
Comment (Leavrth) [inline br/pkg/restore/split/client.go:724]: finally, there might be returning (0, retErr).

**Review by Leavrth — APPROVED** [review]

**Review by Leavrth — APPROVED** [review]

File: br/pkg/lightning/backend/local/localhelper.go [hunk 2]
```diff
@@ -307,14 +307,14 @@ func (local *Backend) SplitAndScatterRegionByRanges(
 	}
 
 	startTime := time.Now()
-	scatterCount, err := local.waitForScatterRegions(ctx, scatterRegions)
-	if scatterCount == len(scatterRegions) {
+	notScatterCount, err := local.splitCli.WaitForScatterRegion(ctx, scatterRegions)
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/local/localhelper.go:310]:
```suggestion
	unScatteredCount, err := local.splitCli.WaitForScatterRegion(ctx, scatterRegions)
```

File: br/pkg/restore/split/client.go [hunk 3]
```diff
@@ -76,18 +78,15 @@ type SplitClient interface {
...
-	// if the latest operator is not `scatter-operator`, or its status is SUCCESS,
-	// it's likely that the scatter region operator is finished.
-	//
-	// if the latest operator is `scatter-operator` and its status is TIMEOUT or
-	// CANCEL, the needRescatter is true and the function caller needs to scatter
-	// this region again.
-	IsScatterRegionFinished(ctx context.Context, regionID uint64) (scatterDone bool, needRescatter bool, scatterErr error)
+	// The first return value is always the number of regions that are not finished
+	// scattering no matter what the error is.
+	WaitForScatterRegion(ctx context.Context, regionInfos []*RegionInfo) (notFinished int, err error)
```
Comment (D3Hunter) [inline br/pkg/restore/split/client.go:89]:
```suggestion
	WaitRegionsScattered(ctx context.Context, regionInfos []*RegionInfo) (notFinished int, err error)
```

**Review by D3Hunter — APPROVED** [review]
rest lgtm

**Pull request merged**